### PR TITLE
[RFC] vim-patch:7.4.2008,7.4.2009

### PR DESCRIPTION
**Pull request by jamessan**
#### vim-patch:7.4.2008

Problem:    evalcmd() has a confusing name.
Solution:   Rename to execute().  Make silent optional.  Support a list of
            commands.

https://github.com/vim/vim/commit/79815f1ec77406f2f21a618c053e5793b597db7a

#### vim-patch:7.4.2009

Problem:    Messages test fails.
Solution:   Don't set redir_execute before returning.

https://github.com/vim/vim/commit/ed59aa60d3905f935283727f4a7b33c81a00174b

### CODE REVIEW

**Comment by brcolow** [comment]
LGTM!

File: src/nvim/testdir/test_execute_func.vim [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  echo 'this'
+  let evaled = execute('echo "that"')
+  echo 'theend'
+  redir END
+"  call assert_equal("\nthis\ntheend", redired)
+  call assert_equal("\nthat", evaled)
+
+  call assert_fails('call execute("doesnotexist")', 'E492:')
+  call assert_fails('call execute(3.4)', 'E806:')
+"  call assert_fails('call execute("call NestedRedir()")', 'E930:')
```
Comment (justinmk) [inline src/nvim/testdir/test_execute_func.vim]: Maybe add a comment:

    " Nvim supports execute('... :redir ...'), so this test is intentionally disabled.

File: test/functional/eval/execute_spec.lua [hunk 2]
```diff
@@ -75,17 +75,59 @@ describe('execute()', function()
...
-    feed(':let g:mes = execute("echon 42")<CR>')
-    screen:expect([[
-    ^                                        |
-    {0:~                                       }|
-    {0:~                                       }|
-    {0:~                                       }|
-    :let g:mes = execute("echon 42")        |
-    ]])
-    eq('42', eval('g:mes'))
+  end)
```
Comment (justinmk) [inline test/functional/eval/execute_spec.lua]: If github is showing this correctly, it looks like an empty test now?

File: test/functional/eval/execute_spec.lua [hunk 3]
```diff
@@ -75,17 +75,59 @@ describe('execute()', function()
...
+      eq(0, exc_exec('let g:mes = execute(0.0, "silent!")'))
+      eq('', eval('g:mes'))
+    end)
+
+    it('propagates errors for "" and "silent"', function()
+      local ret
+      ret = exc_exec('call execute(0.0, "")')
+      eq('Vim(call):E806: using Float as a String', ret)
+
+      ret = exc_exec('call execute(v:_null_dict, "silent")')
```
Comment (justinmk) [inline test/functional/eval/execute_spec.lua]: Isn't this a validation error? So it's not testing the codepath where `f_execute()` actually runs commands that themselves produce errors.

File: test/functional/eval/execute_spec.lua (hunk 2 above)
Comment (jamessan) [inline test/functional/eval/execute_spec.lua]: Oops, meant to remove that.  Its functionality was moved into the new assertions.

File: test/functional/eval/execute_spec.lua (hunk 3 above)
Comment (jamessan) [inline test/functional/eval/execute_spec.lua]: Good point.  Will add some more tests where the error comes out of something that is called.

File: test/functional/eval/execute_spec.lua [hunk 4]
```diff
@@ -62,30 +62,87 @@ describe('execute()', function()
...
+      eq('Vim(echo):E714: List required', ret)
+
+      ret = exc_exec('call execute(["echon 42", "echo add(1, 1)"], "")')
+      eq('Vim(echo):E714: List required', ret)
+
+      ret = exc_exec('call execute("echo add(1, 1)", "silent")')
+      eq('Vim(echo):E714: List required', ret)
+
+      ret = exc_exec('call execute(["echon 42", "echo add(1, 1)"], "silent")')
+      eq('Vim(echo):E714: List required', ret)
```
Comment (justinmk) [inline test/functional/eval/execute_spec.lua]: why wasn't the "42" captured?

Comment (jamessan) [inline test/functional/eval/execute_spec.lua]: It would have been if that was being tested.

Comment (jamessan) [inline test/functional/eval/execute_spec.lua]: Oh, except that running inside of a `try/catch` prevents assignment to the variable.

**Comment by jamessan** [comment]
Ok, I think this is good now.

**Review by justinmk — APPROVED** [review]

**Pull request merged**